teps/0101-env-in-pod-template.md

```diff
@@ -578,9 +674,102 @@ of the file, but general guidance is to include at least TEP number in the
-`podTemplate`, pipelineRun `stepTemplate` can overwrite the taskRun and task `stepTemplate`.
+Bring the task `stepTemplate` spec to the taskRuns and pipelineRuns. Similar to `stepTemplate`, 
+pipelineRun `podTemplate` can overwrite the taskRun and taskRun `podTemplate` can overwrite `stepTemplate`.
+
+#### PipelineRun and TaskRun API Changes
+
+In the `podTemplate` strcut, add the env variable field like:
+
+```go
+type Template struct {
+	...
+	
+	// List of environment variables that can be provided to the containers belonging to the pod.
+	// +optional
+	Envs []corev1.EnvVar `json:"envs,omitempty" patchStrategy:"merge,retainKeys" patchMergeKey:"name" protobuf:"bytes,1,rep,name=envs"`
+	
+	...
+}
+```
+
+`PipelineRun` spec already have `podTemplate` field, so new `Envs` field will be available here
+
+```go
+type PipelineRunSpec struct {
+	....
+	// PodTemplate holds pod specific configuration
+	PodTemplate *PodTemplate `json:"podTemplate,omitempty"`
+    ....
+}
+```
+
+Also, `TaskRun` spec have `podTemplate` field, so `Envs` will be available here
+
+```go
+type TaskRunSpec struct {
+	...
+	
+	// PodTemplate holds pod specific configuration
+	PodTemplate *PodTemplate `json:"podTemplate,omitempty"`
+	
+	...
+}
+```
 
+#### Global Default Environment Variables
+
+Global level default can be provided at the `config-default` configmap under the `default-pod-template` field and the 
+same `podTemplate` field is getting used to read the default values.
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: config-defaults
+  namespace: tekton-pipelines
+data:
+  default-pod-template: | 
+    envs:
+    - 
+    -
+    -
+    ...
+```
+
+#### Precedence Order
+
+We need to have the controller logic to act based on the envs available at different places. Controller should read all
+and merge them till list stage which is converting them to container. The order of Precedence will look like this with 
+first being the higher.
+
+    a. Environment Variables in PipelineRun Pod Template or TaskRun Pod Template (Same priority as both are different places)
+    c. Global Level Default Environment Variables in Tekton Default Pod Template
+    d. Task Level Environment Variables in Task Step Template
+    e. Step Level Environment Variables in Step Container Spec
+
+If some default is configured and user is providing as the time of `PipelineRun` and `TaskRun`, then it should overwrite 
+the default provided at global level.
+
+#### Forbidden Environment Variables
+
+There should be way to define the forbidden environment variables, which should not be overwritten by `PipelineRun` or 
+`TaskRun`. If done so, webhook will throw an error that these are forbidden envs.
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: config-defaults
+  namespace: tekton-pipelines
+  labels:
+    app.kubernetes.io/instance: default
+    app.kubernetes.io/part-of: tekton-pipelines
+data:
+  pod-template-rules: | {
```

Comment:
does it need to be a json inside yaml?

Reply:
yay, this needs to be improved, I will update it. we should have only yaml

Reply:
fixed